### PR TITLE
Replicator filters

### PR DESCRIPTION
**Pull request by Sandychuang8**
Replicator filters: add replication filter interface, add logic to pass document object and isDeleted flag information to client and receive Boolean value to determine to filter the document or not.

### CODE REVIEW

File: android/CouchbaseLite/src/ce/java/com/couchbase/lite/ReplicatorConfiguration.java [hunk 1]
```diff
@@ -213,6 +217,34 @@ public ReplicatorConfiguration setDocumentIDs(List<String> documentIDs) {
         return this;
     }
 
+    /**
+     * Sets a filter closure for validating whether the documents can be pushed
```
Comment (pasin) [inline android/CouchbaseLite/src/ce/java/com/couchbase/lite/ReplicatorConfiguration.java]: Maybe just Sets a filter object?

File: shared/src/main/java/com/couchbase/lite/AbstractReplicator.java [hunk 2]
```diff
@@ -557,6 +565,22 @@ private void _start() {
         boolean pull = isPull(config.getReplicatorType());
         boolean continuous = config.isContinuous();
 
+        C4ReplicationFilter filter = new C4ReplicationFilter() {
+            @Override
+            public boolean validationFunction(final String docID, final int flags, final long dict, final boolean isPush, final Object context)
+                throws  Exception {
+                final AbstractReplicator replicator = (AbstractReplicator) context;
+                return handler.submit(new Callable<Boolean>() {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/AbstractReplicator.java]: I don't think we need to submit the validationFunction call to the looper. We can just call `replicator.validationFunction(docID, flags, dict, isPush)`?

**Comment by pasin** [comment]
Do you have unit tests?

**Comment by Sandychuang8** [comment]
Yes Pasin, all replication related unit tests are in ee, Local2LocalReplicatorTest.java.

File: android/CouchbaseLite/src/ce/java/com/couchbase/lite/ReplicatorConfiguration.java [hunk 3]
```diff
@@ -284,6 +316,18 @@ public Authenticator getAuthenticator() {
         return documentIDs;
     }
 
+    /**
+     * Gets a filter closure for validating whether the documents can be pushed
```
Comment (pasin) [inline android/CouchbaseLite/src/ce/java/com/couchbase/lite/ReplicatorConfiguration.java]: On Java platform, the ReplicationFilter is not a closure. I think use filter object is better.

File: android/CouchbaseLite/src/ce/java/com/couchbase/lite/ReplicatorConfiguration.java [hunk 4]
```diff
@@ -284,6 +316,18 @@ public Authenticator getAuthenticator() {
...
+     * to the remote endpoint.
+     */
+    public ReplicationFilter getPushFilter() { return pushFilter; }
+
+    /**
+     * Gets a filter closure for validating whether the documents can be pulled from the
```
Comment (pasin) [inline android/CouchbaseLite/src/ce/java/com/couchbase/lite/ReplicatorConfiguration.java]: Same as above, I think use filter object is better.

File: android/CouchbaseLite/src/ce/java/com/couchbase/lite/ReplicatorConfiguration.java [hunk 5]
```diff
@@ -284,6 +316,18 @@ public Authenticator getAuthenticator() {
...
+     * remote endpoint. Only documents for which the closure returns true are replicated.
```
Comment (pasin) [inline android/CouchbaseLite/src/ce/java/com/couchbase/lite/ReplicatorConfiguration.java]: For this line, I think change from `closure` to `filter`.

File: shared/src/main/java/com/couchbase/lite/AbstractReplicator.java [hunk 6]
```diff
@@ -36,8 +38,12 @@
 import java.util.Locale;
 import java.util.Map;
 import java.util.Set;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/AbstractReplicator.java]: Can you double check if we need these imports?

File: shared/src/main/java/com/couchbase/lite/AbstractReplicator.java [hunk 7]
```diff
@@ -557,6 +565,17 @@ private void _start() {
         boolean pull = isPull(config.getReplicatorType());
         boolean continuous = config.isContinuous();
 
+        C4ReplicationFilter filter = new C4ReplicationFilter() {
```
Comment (pasin) [inline shared/src/main/java/com/couchbase/lite/AbstractReplicator.java]: I think we should check if there is at least a pushFilter or pullFilter in the config before creating the filter object.

**Pull request merged**